ui/src/app/base/components/ActionButton/ActionButton.js

```diff
@@ -1,58 +1,108 @@
-    "p-icon--success-negative": success && appearance === "negative"
-  });
-
-  return loading || success ? <i className={iconClass} /> : children;
-};
+export const LOADER_MIN_DURATION = 400; // minimium duration (ms) loader displays
+export const SUCCESS_DURATION = 2000; // duration (ms) success tick is displayed
 
 const ActionButton = ({
   appearance = "neutral",
   children,
   className,
+  disabled,
+  inline,
   loading,
   success,
-  width,
   ...props
 }) => {
-  const classes = classNames(className, `p-action-button`, {
-    "is-loading": loading,
-    "is-success": success
-  });
-  const buttonChildren = generateButtonChildren(
-    children,
-    appearance,
-    loading,
-    success
+  const [width, setWidth] = useState(0);
+  const [showLoader, setShowLoader] = useState(false);
+  const [showSuccess, setShowSuccess] = useState(false);
+  const ref = useRef(null);
+
+  // Store default button width
+  useLayoutEffect(() => {
+    if (ref.current && ref.current.getBoundingClientRect().width) {
+      setWidth(ref.current.getBoundingClientRect().width);
+    }
+  }, [children]);
+
+  // Set up loader timer
+  useEffect(() => {
+    let loaderTimeout;
+
+    if (loading) {
+      setShowLoader(true);
+    }
+
+    if (!loading && showLoader) {
+      loaderTimeout = setTimeout(() => {
+        setShowLoader(false);
+
+        if (success) {
+          setShowSuccess(true);
+        }
+      }, LOADER_MIN_DURATION);
+    }
+
+    return () => clearTimeout(loaderTimeout);
+  }, [loading, showLoader, success]);
+
+  // Set up success timer
+  useEffect(() => {
+    let successTimeout;
+
+    if (showSuccess) {
+      successTimeout = setTimeout(() => {
+        setShowSuccess(false);
+      }, SUCCESS_DURATION);
+    }
+
+    return () => clearTimeout(successTimeout);
+  }, [showSuccess]);
+
+  const buttonClasses = classNames(
+    classNames,
+    "p-action-button",
+    `p-button--${appearance}`,
+    {
+      "is-loading": showLoader,
+      "is-success": showSuccess,
+      "is-disabled": disabled,
+      "is-inline": inline
+    }
   );
-  const style = width ? { width } : {};
+
+  const iconClasses = classNames({
+    "p-icon--spinner u-animation--spin": showLoader,
+    "is-light": appearance === "positive" || appearance === "negative",
+    "p-icon--success":
+      showSuccess && (appearance !== "positive" && appearance !== "negative"),
+    "p-icon--success-positive": showSuccess && appearance === "positive",
+    "p-icon--success-negative": showSuccess && appearance === "negative"
+  });
 
   return (
-    <Button
-      appearance={appearance}
-      className={classes}
-      style={style}
+    <button
+      className={buttonClasses}
+      disabled={disabled}
+      ref={ref}
+      style={width ? { width: `${width}px` } : undefined}
       {...props}
     >
-      {buttonChildren}
-    </Button>
+      {showLoader || showSuccess ? <i className={iconClasses} /> : children}
```

Comment:
In the design it shows the text and the spinner at the same time.

Reply:
@huwshimi oh yeah we did try that at first, but the changing text made the button width jump around so we opted to just show a spinner / success icon instead